{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "#Note: code probably needs to be executed twice\n",
    "print (\"Germplasm.tsv\") #A header to make it easier to read\n",
    "with open(\"Germplasm.tsv\") as csvfile: #opening the file\n",
    "    germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"') \n",
    "    #delimiter is a tab, quote character \" in case there is any quote\n",
    "    for entry in germplasm: #For every entry, print the entry as an ordered dictionary\n",
    "        print(entry)\n",
    "        \n",
    "#the same with the other file\n",
    "print (\"LocusGene.tsv\")        \n",
    "with open(\"LocusGene.tsv\") as csvfile:\n",
    "    locusgene = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    for entry in locusgene:\n",
    "        print(entry)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running that code once, we can see that: \n",
    "\n",
    "a) the columns in Germplasm.tsv are \"Locus\", \"germplasm\", \"phenotype\" and \"pubmed\" whereas in \"LocusGene.tsv\" the columns are \"Locus\", \"Gene\" and \"ProteinLength\" (which will be useful for building the database later on- notice use of capital letters and camelcase in the second file)\n",
    "\n",
    "b the the AGI Locus Code in both is filed under the name 'Locus'. That way, instead of printing out the dictionaries for each entry, we can ask especifically to print the AGI Locus Code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "#the same as before\n",
    "print (\"AGI Locus Codes for Germplasm.tsv\")\n",
    "with open(\"Germplasm.tsv\") as csvfile:\n",
    "    germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    for entry in germplasm:\n",
    "        print(entry[\"Locus\"]) #this time, print only the data under the header 'locus'\n",
    "print (\"AGI Locus Codes for Locusgene.tsv\") \n",
    "\n",
    "with open(\"LocusGene.tsv\") as csvfile:\n",
    "    locusgene = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "    for entry in locusgene:\n",
    "        print(entry[\"Locus\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can, therefore, see the list of AGI codes contained in both files and check that the numbers are the same. They are. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 2: Design and create the database.\n",
    "\n",
    "    It should have two tables - one for each of the two data files.\n",
    "    The two tables should be linked in a 1:1 relationship\n",
    "    you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The table design will be as follows:\n",
    "\n",
    "\n",
    "Table 1: Germplasm\n",
    "- Locus (VARCHAR(20), NOT NULL, PRIMARY KEY)\n",
    "- Germplasm (VARCHAR (20), NOT NULL)\n",
    "- Phenotype (VARCHAR (2000), NOT NULL)\n",
    "* the  'phenotype' column contains brief descriptions and so will need more characters of space\n",
    "- Pubmed ID. (INTEGER, NOT NULL)\n",
    "\n",
    "Table 2: Gene\n",
    "- Locus (VARCHAR(20), NOT NULL, PRIMARY KEY)\n",
    "- Gene (VARCHAR (10), NOT NULL)\n",
    "* genes usually are of lesser length but just to ensure there will be no errors\n",
    "- Proteinlength (INTEGER, NOT NULL)\n",
    "\n",
    "Both tables will be linked by the AGI Locus code, which will act as the key, so there is no need to add an additional index ID in the database that might serve as a primary key. It will be added for convenience, and it will have to be coded as a primary key since sql requires auto_increment integers to be primary keys, but for utility purposes the main linkage will be the AGI Locus code.\n",
    "\n",
    "The database will be created using sqlMagic:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " #loading mysql\n",
    "%load_ext sql\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql drop database plantgenes; \n",
    "#this command is only because I've needed to alter parameters and restart the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database plantgenes;\n",
    "%sql use plantgenes;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE germplasm (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(30) NOT NULL,  germplasm VARCHAR(50) NOT NULL, phenotype VARCHAR(2000) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql CREATE TABLE gene (id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, locus VARCHAR(30) NOT NULL, gene VARCHAR(10) NOT NULL, protein_length INTEGER NOT NULL);\n",
    "%sql SHOW TABLES;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Adding the Germplasm table from the file\n",
    "\n",
    "import pymysql.cursors \n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='plantgenes',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open (\"Germplasm.tsv\") as csvfile:\n",
    "            germplasm = csv.DictReader (csvfile, delimiter=\"\\t\", fieldnames=(\"Locus\", \"germplasm\", \"phenotype\", \"pubmed\"))\n",
    "            for entry in germplasm:\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed) VALUES (\"{}\", \"{}\", \"{}\", {})\"\"\".format(entry[\"Locus\"], entry[\"germplasm\"], entry[\"phenotype\"], entry[\"pubmed\"])\n",
    "                cursor.execute(sql) \n",
    "            connection.commit()   \n",
    "            \n",
    "finally:\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%sql SELECT * from germplasm;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data has been successfully added. Since the header also got added into the list as id=1, all we need to do to curate the database is delete it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql DELETE FROM germplasm where id=1;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this block is not strictly needed since I already did it before\n",
    "import pymysql.cursors \n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='plantgenes',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open (\"LocusGene.tsv\") as csvfile:\n",
    "            locusgene = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"')\n",
    "            for entry in locusgene:\n",
    "                print (entry) #control to see if the data is being read correctly\n",
    "                sql = \"\"\"INSERT INTO gene (Locus, gene, protein_length) \n",
    "                VALUES (\"{}\", \"{}\", {})\"\"\".format(entry[\"Locus\"], entry[\"Gene\"], entry[\"ProteinLength\"])\n",
    "                #command for sql to add the data from each column to the database\n",
    "                cursor.execute(sql) \n",
    "            connection.commit() #forcing commit\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * from gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data was successfully added at first, but when I changed the tables to include index id, I started getting errors. I didn't touch anything else."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Problem 4: Create reports, written to a file\n",
    "\n",
    "    Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "    Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "    Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "    Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule!\n",
    "\n",
    "All reports should be written to the same file. You may name the file anything you wish.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Full, joined content of the two database tables\n",
    "#Since connection wasn't closed in Problem 3, we do not need to establish connection again\n",
    "%sql SELECT * from germplasm,gene WHERE gene.locus=germplasm.locus\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is assuming there is no missing gene or germplasm data (which, looking at the table we got with select * from gene, there seems to be none). In case there was missing data, a FULL JOIN or a RIGHT/LEFT join might be needed. Now that we know it works, we can create the script to turn it into a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#importing and establishing a connection is not needed since we did it previously\n",
    "#but added just in case\n",
    "import pymysql.cursors \n",
    "import csv\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='plantgenes',\n",
    "                             charset='utf8mb4',  \n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "try:       \n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"SELECT * from germplasm,gene WHERE gene.locus=germplasm.locus\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #catch results\n",
    "        report = open(\"Report.txt\", \"w\")\n",
    "        report.write(\"Joined content of the two databases:\")\n",
    "        report.write(str(results)) #since write has to be a string, not a list\n",
    "        report.close()\n",
    "finally:\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * from germplasm,gene WHERE gene.locus=germplasm.locus \\\n",
    "AND (gene.gene=\"SKOR\" OR gene.gene=\"MAA3\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, to write this into the file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:       \n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT * from germplasm,gene WHERE gene.locus=germplasm.locus AND (gene.gene=\"SKOR\" OR gene.gene=\"MAA3\")\"\"\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #catch results\n",
    "        report = open(\"Report.txt\", \"a\") #append, since the file was previously created\n",
    "        report.write(\"Genes SKOR and MAA3\")\n",
    "        report.write(str(results)) #since write has to be a string, not a list\n",
    "        report.close()\n",
    "finally:\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For reports 3 and 4, the index ID becomes very useful: We can see that, in the gene table (the germplasm table has different IDs because one entry was deleted - this is however not very important since the linkage is made with AGI Locus Codes) Chromosome 1 spans entries 1-4, Chromosome 2 entries 5-8, Chromosome 3 entries 9-17, Chr4 entries 18-25 and Chromosome 5 entries 26-32. This was counted manually, but a piece of code can also be written for that purpose:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND gene.id<5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:       \n",
    "    with connection.cursor() as cursor:\n",
    "        sql = \"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch1\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND gene.id<5\"\"\"\n",
    "        cursor.execute(sql)\n",
    "        results = cursor.fetchall() #catch results\n",
    "        report = open(\"Report.tsv\", \"a\") #append, since the file was previously created\n",
    "        report.write(str(results)) #since write has to be a string, not a list\n",
    "        report.close()\n",
    "finally:\n",
    "    print(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To not make duplicates (DRY rule), most of the script could be turned into a function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_report (command):\n",
    "    try:       \n",
    "        with connection.cursor() as cursor:\n",
    "            sql = command\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall() #catch results\n",
    "            report = open(\"Report.tsv\", \"a\") #append, since the file was previously created\n",
    "            report.write(str(results)) #since write has to be a string, not a list\n",
    "            report.close()\n",
    "    finally:\n",
    "        print(\"\")\n",
    "\n",
    "add_report(\"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch1\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND gene.id<5\"\"\")\n",
    "add_report(\"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch2\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND 4<gene.id<9\"\"\")\n",
    "add_report(\"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch3\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND 8<gene.id<18\"\"\")\n",
    "add_report(\"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch4\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND 17<gene.id<26\"\"\")\n",
    "add_report(\"\"\"SELECT COUNT(*) AS \"Number Of Matches in Ch5\" FROM gene, germplasm WHERE germplasm.locus = gene.locus AND 25<gene.id<33\"\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The piece of code could have probably incorportated more of the sql command into the function so that the user needed to type less, but since I will also be using a similar function in report 4 but the command will be slightly different (AVG() instead of COUNT()), I found this way to be easy, yet save me from having to write another function. Also, in this case there is no need for the query to span two tables since the information is only found in the gene table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "add_report(\"\"\"SELECT AVG(protein_length) AS \"Average protein length in Ch1\" FROM gene WHERE gene.id<5\"\"\")\n",
    "add_report(\"\"\"SELECT AVG(protein_length) AS \"Average protein length in Ch2\" FROM gene WHERE 4<gene.id<9\"\"\")\n",
    "add_report(\"\"\"SELECT AVG(protein_length) AS \"Average protein length in Ch3\" FROM gene WHERE 8<gene.id<18\"\"\")\n",
    "add_report(\"\"\"SELECT AVG(protein_length) AS \"Average protein length in Ch4\" FROM gene WHERE 17<gene.id<26\"\"\")\n",
    "add_report(\"\"\"SELECT AVG(protein_length) AS \"Average protein length in Ch5\" FROM gene WHERE 25<gene.id<33\"\"\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "connection.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
